### ISSUE TITLE
Opset Coverage - Binary Size Tradeoff

### ISSUE THREAD

**Issue by adityagoel4512** [microsoft/onnxruntime#15397]
### Discussed in https://github.com/microsoft/onnxruntime/discussions/15353

<div type='discussions-op-text'>

<sup>Originally posted by **adityagoel4512** April  4, 2023</sup>
We express models and feature engineering inference pipelines our clients have developed using ONNX. As it stands, the policy to only include operators (and specifically, support all argument types supported by the standard for a given operator) as and when needed out of concern for binary size can significantly slow down iteration speed and make user experience significantly worse since you only find out ORT doesn't support something you need (and cannot work around) when you need it most!

Given custom build options exist for [reduced operator builds](https://onnxruntime.ai/docs/build/custom.html#build-configuration), users already have a straightforward way to build onnxruntime with a smaller binary footprint should they need this. 

I would make the argument that _aiming_ to support the entire standard and giving users optionality to build ORT with a smaller binary footprint strikes the right balance. We currently cannot specify reduced type builds where we only include certain types supported by an operator so this might be some further tooling that could be supported.

I would be interested to hear any thoughts the community has.

</div>

**Comment by yuslepukhin** [comment]
Cc: @skottmckay @edgchen1

**Comment by pranavsharma** [comment]
@adityagoel4512 Not quite following. We already have what you're asking for.

> I would make the argument that aiming to support the entire standard and giving users optionality to build ORT with a smaller binary footprint strikes the right balance.

Our default build supports the whole standard. We intentionally don't implement types that are not required for production scenarios at all. You're always welcome add those types if there's a real use case.

> We currently cannot specify reduced type builds where we only include certain types supported by an operator

Have you seen this? https://onnxruntime.ai/docs/reference/operators/reduced-operator-config-file.html

**Comment by adityagoel4512** [comment]
> @adityagoel4512 Not quite following. We already have what you're asking for.
> 
> > I would make the argument that aiming to support the entire standard and giving users optionality to build ORT with a smaller binary footprint strikes the right balance.
> 
> Our default build supports the whole standard. We intentionally don't implement types that are not required for production scenarios at all. You're always welcome add those types if there's a real use case.
> 
> > We currently cannot specify reduced type builds where we only include certain types supported by an operator
> 
> Have you seen this? https://onnxruntime.ai/docs/reference/operators/reduced-operator-config-file.html

My point was more to do with this decision to intentionally not cover all the types until someone discovers a production scenario:

> We intentionally don't implement types that are not required for production scenarios at all. You're always welcome add those types if there's a real use case.

While the PR based model to add the types as and when needed does certainly work, I wanted to know what the reason might be behind this is? I imagine it's binary size?

**Comment by pranavsharma** [comment]
Couple of reasons: 
1. We don't want our default binary size to grow unnecessarily. While users can always restrict ops and types in their custom builds it's not the most convenient experience and intended for only those users who are really constrained. It would be unfair to penalize our regular users by increasing the size without any bounds just to support the standard and impose artificial constraints on them forcing them to resort to custom builds.
2. Adding support for certain types (e.g., complex types) is not only non-trivial taking resources away from higher priority work items but also adds to the complexity of the codebase and increases maintenance costs.

Having said that if there's a legitimate production use case we're more than happy to accept contributions or make the changes ourselves when time permits.

**Comment by skottmckay** [comment]
> I would make the argument that _aiming_ to support the entire standard and giving users optionality to build ORT with a smaller binary footprint strikes the right balance.

A custom build involves a significant amount of time and effort, especially if you're deploying on multiple platforms and/or architectures. For example, if you're using the nuget package you have to assemble builds from all platforms in one place before creating the package, which is non-trivial. 

Our current pre-built binaries include the necessary types to handle known production scenarios. Occasionally a new unsupported scenario comes up and we add the required types to the necessary operators. Historically this has been infrequent. Wouldn't the development/maintenance/binary size cost to support all types in order to avoid these infrequent type expansions be optimizing for an edge case?

**Comment by cbourjau** [comment]
I feel as if there is a bit of ambiguity about which types and binary sizes/EPs are referred to here. Maybe it helps to be more explicit here?

Our ( @adityagoel4512 and I are colleagues) primary use case at this point is expressing feature engineering pipelines in ONNX. We want to empower data scientists (on our side and on the client side) to express their business logic of that kind in ONNX. The pipelines have to be as close as possible to their numpy/pandas equivalent. We are exclusively running these models on the CPU execution provider. Numpy/Pandas defaults to `float64`, `int64`, and `object`/`string` making these two data types particularly important to us. To a lesser degree, we may also encounter `float32`, `int32` and `uint64` data types. 

In summary, having good and predictable coverage of the standard for the already existing data types (`float64`, `int64`, `string`, `float32, `int32`, ...) on the CPU execution provider would be very helpful to us. Our use case currently does not call for expanding the coverage for other execution providers or introducing entirely new types (such as complex ones). 

The binary size increase required to implement the above seems fairly marginal especially since it is  GPU rather than CPU releases that are pretty heavyweight in terms of size.

Given these clarifications of our use case, is there interest from the `onnxruntime` project side to accept PRs that expand the coverage for these types on the CPU EP proactively or do you consider our use case of expressing Numpy/Pandas feature engineering pipelines in ONNX too much of an edge case? Maybe the ORT could select a few "blessed" data types for which it strives to have complete coverage while non-blessed data types continue to be implemented on a per-use-case basis?

**Comment by skottmckay** [comment]
> The binary size increase required to implement the above seems fairly marginal especially since it is GPU rather than CPU releases that are pretty heavyweight in terms of size.

'marginal' is relative. The impact of the binary size increase on mobile/edge devices could very well be significant.

Whilst numpy/pandas might default to 64-bit numbers, producing models that use them is roughly twice as expensive in terms of model size and memory usage, and incurs the performance penalty of processing 2x the data vs a 32-bit version of the model. Unless the use-case requires 64-bit numbers, that is a high price to pay. e.g. I'd expect that roughly doubles the cost of training the model. 

AFAIK there should be complete coverage for the common types - e.g. float32 data/weights, uint8/int8 in a quantized model, and int64 when used for indices. We have added to individual operators the types required by all production scenarios that we are aware of over a number of years, so based on that I'd expect all common use cases should be supported. 

Using unnecessarily large data types probably falls into an uncommon use case though given the high cost involved. 

Is there a reason not to convert to 32-bit as part of your pipeline?

**Comment by cbourjau** [comment]
I am aware of the benefits of of using `float32`. I think our use case is novel in the sense that we are really striving to express long and complex feature engineering steps completely in ONNX which have previously been expressed in Pandas/Numpy. This is probably the reason why we have run into missing `float64` implementations repeatedly. Would it be better to express the pipeline in `float32` even in Pandas/Numpy? Often times yes! But that is not how pandas works. Memory and speed is not the primary concern in many use cases and thus models end up being trained using `float64` precision. If we convert the models using `float32` we get discrepancies that need to be monitored and explained. This is especially important when using decision trees where a loss of precision can lead to ending up in a completely different branch. This may sound unlikely but it happens in reality. For instance, LightGBM's algorithm for determining split values can lead to splits coinciding *precisely* with values seen in the training data in some scenarios. This is very unfortunate but the reality. Using `float32` for the feature engineering at inference time will lead to completely different outputs. 

**tl;dr**:
- Pandas/Numpy defaults to `float64`. This behavior is hard to control and leads to (legacy) models being trained with that precision.
- Supporting `float64` makes the introduction of ONNX into an organization much easier because it removes any discussions in the review due to numerical discrepancies. Some use cases absolutely require `float64` in the feature engineering pipeline.
